### ISSUE TITLE
Automatically update Cypress files with VA.gov user viewport data once a month

### ISSUE THREAD

**Issue by holdenhinkle** [department-of-veterans-affairs/va.gov-team#16994]
## Issue Description
Instead of manually logging into the VA.gov Google Analytics account once a month to manually get the latest viewport information for mobile, tablet and desktop devices, then updating the code, and finally submitting a PR, we'd like to automate this process.

---
## Tasks
- [x] Create a Sidekiq(?) job that executes a script once a month, on the first of every month
- [x] The script should get the the following data for the mobile, tablet and desktop devices: viewport width and height, percent traffic which uses the viewport
- [x] Script should update the `config/cypress.json` and `src/platform/testing/e2e/cypress/support/commands/viewportPreset.js` files (or something like that)
- [x] The script should submit a PR for the updated files (or something like that)

## Acceptance Criteria
- [x] A new Sidekiq(?) job will be created, along with a script that is executed once a month that updates the `cypress.json` file with last month's viewport data

**Comment by holdenhinkle** [comment]
Google doc describing, in detail, steps to build this script - https://docs.google.com/document/d/1GQOxJxxtE2xt43pQxsbO9a-bxiaDKDPheOSoKL7JjbA/edit?usp=sharing

Major questions:

- What background jobs system should we use?
  - BE uses sidekiq
  - Is there another background jobs system I should be aware of?
- Who do I speak with about registering this application in the Google API Console so I can get the client ID and client secret?
- Who do I contact about registering this application with GitHub?

**Comment by holdenhinkle** [comment]
I had a conversation with #vsp-analytics about getting access to GA - https://dsva.slack.com/archives/CKCRMLGAC/p1610049987173200

I filled out a request form for GA API access - https://github.com/department-of-veterans-affairs/va.gov-team/issues/18267

I started a conversation about getting Github API access - https://dsva.slack.com/archives/C01CHAY3ULR/p1610050405487600

**Comment by holdenhinkle** [comment]
UPDATE

I received received the Google api credentials I need to hit the Google Analytics Reports API (`project_id`, `private_key_id`, `private_key`, etc.) The credentials were sent to me via my Keybase account by Brian Martin in #vsp-analytics. Here's the thread - https://dsva.slack.com/archives/CKCRMLGAC/p1610049987173200

I submitted a request to github-user-requests for them to create a GitHub App for me. Here's the request - https://github.com/department-of-veterans-affairs/github-user-requests/issues/5742

I spent a few hours learning about the different aways to interact with GitHub's API (using OAuth or GitHub Apps) and determined that GitHub Apps is the way to go, mainly because:

1) GitHub recommends GitHub Apps for use cases like this
2) the app will appear as a bot that's making the commit, submitting the PR, etc. instead of a user
3) GitHub says there's improved security in Apps vs creating a user account for things like this and using OAuth

Some of the resources I used:

- Video: https://www.youtube.com/watch?v=iaBEWB1As0k
- Slides from the Video: https://gist.github.com/swinton/96d4d12c06c2c2b1256ffb0207acf9ba
- GitHub App Cheatsheet: https://github.com/github-developer/github-apps-cheat-sheet/blob/master/README.md 
- How to create a GitHub App: https://docs.github.com/en/free-pro-team@latest/developers/apps/creating-a-github-app
- Octokit - Ruby toolkit for the GitHub API: https://github.com/octokit/octokit.rb

I reviewed the GitHub API documentation and made sure the parts we need to use (create branch, make commit, submit pr) works with GitHub Apps:

Create PR - Works with GitHub Apps ✅
https://docs.github.com/en/free-pro-team@latest/rest/reference/pulls#create-a-pull-request

Create a reference (branch) - Works with GitHub Apps ✅
https://docs.github.com/en/free-pro-team@latest/rest/reference/git#create-a-reference 

Create a commit - Works with GitHub Apps ✅
https://docs.github.com/en/free-pro-team@latest/rest/reference/git#create-a-commit

I chatted with Lindsey Hattamer about creating the app in `vets-api` so I can configure Sidekiq to run this job each month. I asked her if I need any kind of permission to create a new Sidekiq job and she said, 'Nope, no permission needed for a new job!'.

I updated the Google Doc with more notes and resources - https://docs.google.com/document/d/1GQOxJxxtE2xt43pQxsbO9a-bxiaDKDPheOSoKL7JjbA/edit#

I'll start coding this on Monday!

**Comment by holdenhinkle** [comment]
Response from "github-user-requests" regarding request to create a GitHub App - https://github.com/department-of-veterans-affairs/github-user-requests/issues/5742#issuecomment-757483751

> Hey @holdenhinkle!
> 
> Easiest way to do this is for you to create the GitHub App against your account, then request a transfer into the department-of-veterans-affairs organization.
> 
> https://docs.github.com/en/free-pro-team@latest/developers/apps/transferring-ownership-of-a-github-app

**Comment by holdenhinkle** [comment]
Branch - https://github.com/department-of-veterans-affairs/vets-api/tree/16994_update_vets_website_config/cypress.json_file

**Comment by holdenhinkle** [comment]
Completed today:

Created two Google Analytics reports using Google API:
1. total users for previous month
2. total users for each viewport for previous month, sorted by users descending, limit 100

Total users is parsed from the first report which is need to calculate the percentage of users who use each viewport.

A viewport collection is created from the second report for top mobile, tablet, and desktop viewports. The viewports are hashes that match the structure of the viewports objects in cypress.json

Next:
- get cypress file
- edit file
- save file
- create new git branch
- commit file
- submit pr for updated file

**Comment by holdenhinkle** [comment]
Created GiHub app and transferred it to department-of-veterans-affairs org. Awaiting approval...

**Comment by holdenhinkle** [comment]
I have to update this file too - https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/testing/e2e/cypress/support/commands/viewportPreset.js

The presets must be updated.

**Comment by holdenhinkle** [comment]
Requested new user be created for this worker - https://github.com/department-of-veterans-affairs/github-user-requests/issues/5868

Tasks remaining:
- Write unit tests
- Schedule the job

**Comment by holdenhinkle** [comment]
Status Update:

Lindsey H. asked that the main PR be broken into parts, the first bing a pr to add the new gems:

> Lindsey Hattamer  12:59 PM
> Actually i'd create a separate PR for the gemfile/gemfile.lock for the new gem you're introducing
> 
> Holden Hinkle  1:35 PM
> Hey, just got back from lunch.
> 
> 1:36
> Okay, I’ll submit a pr for the gems today. I’m probably submit that other (BIG) pr later too. Yeah, it’s a big one. So you would have preferred that to be broken up some how? (edited) 
> 
> Lindsey Hattamer  1:40 PM
> Yes please. I'd try to break it up in whichever way makes the most sense wrt grouping
> 
> 1:00
> I'll be easier to review since we promote smaller PRs :slightly_smiling_face:
> 
> Yes please. I'd try to break it up in whichever way makes the most sense wrt grouping

The main branch has been broken into 7 parts. Part 1 is in draft mode here - https://github.com/department-of-veterans-affairs/vets-api/pull/5800

Still blocked on setting up a new 'bot' user with access to Github that will provide the required personal access token. Peter Hill is working on this.

Once we get the token, I'll quickly complete the following and submit the 7 prs:
- Add Google and Github credentials to `credstash` (Jeremy is DevOps is going to help me with this)
- Update references to credentials from `settings.local.yml` to `credstash`
- Update references to https://github.com/holdenhinkle/vets-website in `spec` files
- Delete the GitHub VCR cassettes and re-record the requests/responses to https://github.com/department-of-veterans-affairs/vets-website

**Comment by pjhill** [comment]
* VSP Testing Tools Team mailing list created
* VSP Testing Tools Team mailing list GitHub account has been created
* Credentials for VSP TTT mailing list stored in [va.gov-team-sensitive](https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/blob/master/platform/quality-assurance/vsp-testing-tools-team.md)

**Comment by holdenhinkle** [comment]
Status Update:

Got this working with the GitHub App instead of the new GitHub user + personal access token. When I got the new user and token yesterday afternoon the updater didn't work because it couldn't write to GitHub. After confirming that everything should be working, I paired with Peter. Peter mentioned that the login to `department-of-veterans-affairs` is protected by 2fa. I instantly new this was the problem. After researching two factor authentication using the non-OAuth flow + personal access token I concluded that 2fa + personal access token isn't possible. (It's possible using the GitHub command line but that's an app.)

I circled back to GitHub Apps this morning and cracked around 11:30. Everything is working beautifully and all tests pass.

Because I switched from GitHub user + personal access token to GitHub apps, I looked at VCR Cassette recordings to make sure no new sensitive data is being recorded. I noticed that updated file requests (put requests) had a PGP SIGNATURE in the response body. Peter and I researched whether or not it's safe to leave them in the VCR Cassette yml files but couldn't tell for sure. Peter suggested that I check in with Joe Niquette about this which I did. Joe's looking into it. He's pretty sure it's safe but is waiting to review the files I sent him through Keybase (which didn't go through the first time around).

I worked with Lindsey H in BE Tools and Jeremy in Devops about where to store the Google Analytics and GitHub settings. I got a great lesson on how and where to store the settings. I submitted a PR to update the settings.local.yml file for the dev, staging, prod and sandbox environments - https://github.com/department-of-veterans-affairs/devops/pull/8457#issuecomment-772048764 Jeremy said he'd add the settings to credstash first thing tomorrow morning then approve the pr.

I made various commits to the main branch, changing out how settings are referenced, updating tests now that the updater is working with the `department-of-veterans-affairs/vets-api` repo instead of `holdenhinkle/vets-api`, etc. - https://github.com/department-of-veterans-affairs/vets-api/compare/16994_update_vets_website_config/cypress.json_file

Once credstash and the settings.local.yml files are updated tomorrow morning and Joe confirms that it's okay to commit those VCR cassette files I'll update all the branches for the updater and finally submit all the PRs.

**Comment by holdenhinkle** [comment]
The following PRs where submitted on the 2nd and 3rd of Feb, 2021:

add `google_analytics_cvu` and `github_cvu` settings to `dev`, `staging`, `prod`, and `sandbox` `settings.local.yml` files -  https://github.com/department-of-veterans-affairs/devops/pull/8457  (STATUS: MERGED)

Cypress Viewport Updater Bot app
- Part 1 - https://github.com/department-of-veterans-affairs/vets-api/pull/5800 (STATUS: MERGED)
- Part 1.5 - https://github.com/department-of-veterans-affairs/vets-api/pull/5840 (STATUS: APPROVED)
- Part 2 - https://github.com/department-of-veterans-affairs/vets-api/pull/5841
- Part 3 - https://github.com/department-of-veterans-affairs/vets-api/pull/5842
- Part 4 - https://github.com/department-of-veterans-affairs/vets-api/pull/5845
- Part 5 - https://github.com/department-of-veterans-affairs/vets-api/pull/5846
- Part 6 - https://github.com/department-of-veterans-affairs/vets-api/pull/5847
- Part 7 - https://github.com/department-of-veterans-affairs/vets-api/pull/5848

Cypress Viewport Updater Bot documentation
https://github.com/department-of-veterans-affairs/va.gov-team/pull/19476  (STATUS: MERGED)

**Comment by holdenhinkle** [comment]
All PRs have been merged and the `config/cypress.json` and `src/platform/testing/e2e/cypress/support/commands/viewportPreset.js` files were automatically updated on Saturday (I ran the job manually because it's not set to run again until March).

This ticket is completed.

**Comment by pjhill** [comment]
Hooray -- great job!
